src/flamenco/txn/fd_txn_generate.c

```diff
@@ -0,0 +1,158 @@
+  ulong signers_write_sz = FD_TXN_ACCT_ADDR_SZ * (accounts->signature_cnt - accounts->readonly_signed_cnt);
+  fd_memcpy( write_ptr, accounts->signers_w, signers_write_sz );
+  write_ptr += signers_write_sz;
+
+  fd_memcpy( write_ptr, accounts->signers_r, FD_TXN_ACCT_ADDR_SZ * accounts->readonly_signed_cnt );
+  write_ptr += FD_TXN_ACCT_ADDR_SZ * accounts->readonly_signed_cnt;
+
+  ulong non_signers_write_sz = FD_TXN_ACCT_ADDR_SZ * (ulong)(accounts->acct_cnt - accounts->readonly_unsigned_cnt - accounts->signature_cnt);
+  fd_memcpy( write_ptr, accounts->non_signers_w, non_signers_write_sz);
+  write_ptr += non_signers_write_sz;
+
+  fd_memcpy( write_ptr, accounts->non_signers_r, FD_TXN_ACCT_ADDR_SZ * accounts->readonly_unsigned_cnt );
+  write_ptr += FD_TXN_ACCT_ADDR_SZ * accounts->readonly_unsigned_cnt;
+  FD_TEST( (ushort)((ulong)write_ptr - (ulong)out_txn_payload) == txn_meta->recent_blockhash_off );
+
+  /* Write recent blockhash */
+  if( FD_LIKELY( opt_recent_blockhash ) ) {
+    memcpy( write_ptr, opt_recent_blockhash, FD_TXN_BLOCKHASH_SZ );
+  } else {
+    memset( write_ptr, 0UL, FD_TXN_BLOCKHASH_SZ );
+  }
+  write_ptr += FD_TXN_BLOCKHASH_SZ;
+
+  return (ulong)(write_ptr - out_txn_payload);
+}
+
+ulong
+fd_txn_add_instr( uchar * txn_meta_ptr,
+                  uchar out_txn_payload[ static FD_TXN_MTU ],
+                  uchar program_id,
+                  uchar const * accounts,
+                  ulong accounts_sz,
+                  uchar const * instr_buf,
+                  ulong instr_buf_sz ) {
+
+  fd_txn_t * txn_meta = (fd_txn_t *) txn_meta_ptr;
+  FD_TEST( txn_meta->instr_cnt < FD_TXN_INSTR_MAX );
+  FD_TEST( txn_meta->recent_blockhash_off != 0 );
+
+  uchar * instr_start = out_txn_payload + txn_meta->recent_blockhash_off + FD_TXN_BLOCKHASH_SZ;
+  txn_meta->instr_cnt++;
+  uchar * write_ptr       = instr_start;
+
+  uchar compact_instr_cnt_sz = (uchar) fd_cu16_enc( (ushort)txn_meta->instr_cnt, write_ptr );
+  FD_TEST( compact_instr_cnt_sz == 1 );
+
+  write_ptr += compact_instr_cnt_sz;
+
+  /* Calculate offset of next instruction. */
+  if ( txn_meta->instr_cnt > 1 ) {
+    write_ptr = out_txn_payload + txn_meta->instr[txn_meta->instr_cnt-2].data_off + txn_meta->instr[txn_meta->instr_cnt-2].data_sz;
+  }
+
+  instr_start = write_ptr;
+
+  *write_ptr = program_id;
+  write_ptr += sizeof(uchar);
+
+  uchar compact_accts_len_sz = (uchar) fd_cu16_enc( (ushort)accounts_sz, write_ptr );
+  write_ptr += compact_accts_len_sz;
+
+  fd_memcpy( write_ptr, accounts, accounts_sz );
+  write_ptr += accounts_sz;
+
+  ushort data_sz = (ushort)instr_buf_sz;
+  uint compact_data_len_sz = fd_cu16_enc( data_sz, write_ptr );
+  write_ptr += compact_data_len_sz;
+
+  /* Copy data buffer over */
+  ushort data_off = (ushort) (write_ptr - out_txn_payload);
+  fd_memcpy( write_ptr, instr_buf, data_sz );
+  write_ptr += data_sz;
+  ushort acct_off = (ushort) (instr_start + sizeof(uchar) + compact_accts_len_sz - out_txn_payload);
```

Comment:
Why not just move this line a few lines up and calculate it the same way as `data_off` ?